extensions/amp-story/amp-story-animation.md

```diff
@@ -0,0 +1,648 @@
+You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS-IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# amp-story-animation
+
+The `<amp-story-animation>` element defines a custom animation effect in an
+[`amp-story`](./amp-story.md). These effects rely on the [Web Animations API](https://www.w3.org/TR/web-animations/) to define and run animations in AMP Stories.
+
+## Example
+
+```html
+<amp-story-page id="cover">
+  <amp-story-grid-layer template="fill">
+    <div class="zoom">Zoom!</div>
+  </amp-story-grid-layer>
+  <amp-story-animation layout="nodisplay" trigger="visibility">
+    <script type="application/json">
+      {
+        "selector": ".zoom",
+        "iterations": "infinite",
+        "keyframes": {
+          "transform": ["scale(1)", "scale(1.2)", "scale(1)"]
+        }
+      }
+    </script>
+  </amp-story-animation>
+</amp-story-page>
+```
+
+## Attributes
+
+### trigger
+
+Determines when the animation should be triggered. This must be set to
+`visibility` so the animation starts when a story page becomes visible and active.
+
+### layout
+
+Should always be `nodisplay`.
+
+### id [optional]
+
+An ID can be used to determine [sequencing](./amp-story.md#sequencing-animations)
+when defining multiple animations. Other animated elements or `<amp-story-animation>` effects can use this ID to determine that they should
+start animating only after this effect has finished.
+
+### animate-in-after [optional]
+
+Use this attribute to chain or sequence animations. Specify the ID of the animated element that this element's animation will follow. The element must be present on the same `<amp-story-page>`. For further details, see the [Sequencing animations](./amp-story.md#sequencing-animations) section.
+
+For example, in the following code, the `amp-story-animation` effect starts
+after `object1` completes its entrance:
+
+```html
+<amp-story-page id="page1">
+  <amp-story-grid-layer template="vertical">
+    <div id="object1" animate-in="rotate-in-left">
+      1
+    </div>
+    <div class="custom-animated">
+      2
+    </div>
+  </amp-story-grid-layer>
+  <!-- will start after object1 has finished -->
+  <amp-story-animation
+    layout="nodisplay"
+    trigger="visibility"
+    animate-in-after="object1"
+  >
+    <script type="application/json">
+      {
+        "selector": ".custom-animated",
+        "keyframes": {
+          "opacity": [0, 1],
+          "transform": ["scale(0.5)", "scale(1)"]
+        }
+      }
+    </script>
+  </amp-story-animation>
+</amp-story-page>
+```
+
+## Format
+
+An `amp-story-animation` element defines such an animation as a JSON structure.
```

Comment:
IMHO we should at least link/mention `amp-animation`, explain the relationship, and mention that the format mostly comes from there. It's fine to repeat it here, but we need to explain the relationship between the two, otherwise it's just will be an uncertainty.

Reply:
I'm not sure we need to explain the relationship. `amp-animation` is not valid in AMP stories and `amp-story-animation` is not valid on websites. What is important about the relationship?

Reply:
@alanorozco how much of the documentation do `amp-animation` and `amp-story-animation` share?

Reply:
@CrystalOnScript I had some concerns of someone looking at this and asking "Is amp-story-animation same thing as amp-animation? How are they different?". I do realize that user bases might not necessarily overlap a lot. But still seems like a glaring question to me.

Reply:
Out of curiosity, why didn't we make amp-animation valid for the story platform? What is the difference between the two? I might be able to suggest how to "link" the two with more information.

Reply:
It uses the same formats and engine inside. But we judged that the customizations that Stories apply to amp-animation were still making it different enough to warrant a different name for it. However, the internal format is still the same. For instance, we could add new type of animated property to `amp-animation`. It'd automatically work in the `amp-story-animation`, but will we remember to update both docs?

Reply:
Because of the way amp.dev filters by format, I fear that linking from `amp-story-animation` to `amp-animation` will cause more confusion that it will solve. 

<img width="1680" alt="Screen Shot 2020-05-11 at 9 46 55 PM" src="https://user-images.githubusercontent.com/23297417/81639500-f6b8ac80-93d0-11ea-82fc-18977d24785e.png">

Does `amp-story-animation` work the same as other children of `amp-story` where it only requires the `amp-story` script? Or does it import its own? 

From the developer's perspective, what is different about building an animation with `amp-animation` and `amp-story-animation`?

However, I understand your fear about maintaining two docs. `amp-animation`'s is actually in the middle of a restructure from the vendor team I'm working with. Once I understand more of the differences between the components, I may be able to prioritize the update to best meet the needs of both components. Depending on the change, we may be able to pull documentation from `amp-animation` into `amp-story-animation` during amp.dev's build.

Reply:
> Does amp-story-animation work the same as other children of amp-story where it only requires the amp-story script? Or does it import its own?

It only requires the `amp-story` script.

> From the developer's perspective, what is different about building an animation with amp-animation and amp-story-animation?

1. `amp-story-animation` does not support keyframes from CSS or 
composition.

2. `amp-story-animation` can sequence animations with the `animate-in-after` attribute.

I understand the concerns with duping. I wonder if it would make sense to make the `amp-animation` doc an alias for `amp-story-animation` by toggling sections conditionally based on format?

Reply:
Thanks for chiming in. Give me till EOD to update the `amp-animation` doc with these differences in mind. Calling @sebastianbenz for any ideas on how to best organize this in amp.dev.

Reply:
Ok. I recommend we check this in a timeboxed manner. And if nothing better pops up - we'll merge and follow up when this becomes pressing.

Reply:
I restructured the documentation for `amp-animation` in [PR #28371](https://github.com/ampproject/amphtml/pull/28371). I think the order and structure makes a bit more sense. Let me know what you think!

Unfortunately, it solves none of the documentation problems related to `amp-story-animation`. This one is a doozy. ya'll got me.

We can't filter the `amp-animation` documentation for `amp-story-animation`, as they are, in fact, two separate components and all code samples would be wrong, along with the import script instructions. I am also unable to provide a solution to import relevant documentation from `amp-animation` into `amp-story-animation` quickly enough for this release. 

If we have a high concern about maintaining two sets of docs at this time, it makes sense to link out to the `amp-animation` documentation. To avoid confusion on amp.dev, I would put in hard-coded links for GitHub (for now). We can add a blurb explaining `amp-story-animation` is based off of `amp-animation`, implementation is pretty similar, except you don't import a script and and `amp-story-animation` doesn't support keyframes from CSS or composition - also it exposes the `animate-in-after` attribute! 

I'm happy to try to write something a bit more cohesive once we agree on a plan (even if it is temporary). Also happy to sync up for a chat about possible ideas.

Thanks for your patience.

Reply:
Do you think it'd be bad to move the "Animation definition JSON" to a separate md file and link it from both `amp-animation` and `amp-story-animation`? This is actually how our code is written :)

Reply:
But to emphasize: I think this way is good and we should merge and revisit as needed.

Reply:
Configuring an animation in AMP can be a confusion process. Eventually, we would like to increase the amount of in-line executable code samples so I don't think pulling out the definitions is the right solution. 

For now, I've put together a temporary solution in this doc:
https://docs.google.com/document/d/140DN55-M3ptME89mJTwYCa_Au19Zwync1xOFIQnJJQk/edit?usp=sharing

It does rely on merging the [documentation PR to `amp-animation`](https://github.com/ampproject/amphtml/pull/28371).

Reply:
@CrystalOnScript I did some comments on the doc. But do we need to block this PR by the #28371?

Reply:
I would at least update the `md` file in this PR to align with some of the changes outlined in [PR #28371](https://github.com/ampproject/amphtml/pull/28371) as this documentation does not reflect some of the updates and corrections.